File: balanceUpdater.js
import fs from 'fs/promises';

/**
 * Una clase para obtener los movimientos del tablón de una liga de Biwenger,
 * filtrarlos y agruparlos por usuario.
 */
export class BalanceUpdater {
    /**
     * Inicializa el actualizador de balances.
     * @param {number} leagueId El ID de la liga.
     * @param {number} userId El ID del usuario para la autenticación.
     * @param {string} token El token de autorización (Bearer).
     * @param {string} inputFile Ruta al JSON de equipos.
     * @param {string} outputFile Ruta donde se guardará el JSON resultante.
     */
    constructor(leagueId, userId, token, inputFile, outputFile) {
        this.apiUrl = `https://biwenger.as.com/api/v2/league/${leagueId}/board`;
        this.headers = {
            'x-league': String(leagueId),
            'x-user': String(userId),
            'x-version': '628',
            'Authorization': `Bearer ${token}`
        };
        this.inputFile = inputFile;
        this.outputFile = outputFile;
        this.allowedTypes = new Set(["transfer", "market", "clauseIncrement", "roundFinished", "adminTransfer"]);
    }

    /** Carga los IDs de los equipos desde el fichero JSON de entrada. */
    async _loadTeamIds() {
        try {
            const fileContent = await fs.readFile(this.inputFile, 'utf-8');
            const teams = JSON.parse(fileContent);
            // Usamos un Set para una búsqueda más eficiente
            return new Set(teams.map(team => team.id));
        } catch (error) {
            if (error.code === 'ENOENT') {
                console.error(`Error: El fichero de entrada no fue encontrado en '${this.inputFile}'`);
            } else {
                console.error(`Error al leer o parsear el fichero '${this.inputFile}':`, error);
            }
            return null;
        }
    }

    /** Obtiene todas las transacciones de la API, manejando la paginación. */
    async _fetchAllTransactions() {
        const allTransactions = [];
        let offset = 0;
        const limit = 100;

        console.log("Iniciando la obtención de transacciones de la API...");
        while (true) {
            const url = new URL(this.apiUrl);
            url.searchParams.append('offset', String(offset));
            url.searchParams.append('limit', String(limit));

            try {
                const response = await fetch(url.toString(), { headers: this.headers });
                if (!response.ok) {
                    throw new Error(`Error de la API: ${response.status} ${response.statusText}`);
                }
                const result = await response.json();
                const data = result.data || [];

                if (data.length === 0) {
                    console.log("No se encontraron más transacciones. Finalizando paginación.");
                    break;
                }

                allTransactions.push(...data);
                console.log(`Obtenidas ${data.length} transacciones. Total acumulado: ${allTransactions.length}`);

                if (data.length < limit) break;

                offset += limit;
            } catch (error) {
                console.error(`Error al contactar la API de Biwenger:`, error);
                return null;
            }
        }
        return allTransactions;
    }

    /**
     * Procesa las transacciones para calcular ingresos y gastos, y las agrupa por usuario.
     * @param {Array} transactions - La lista de todas las transacciones.
     * @param {Set<number>} teamIds - Un Set con los IDs de los equipos de la liga.
     * @returns {Object} Un objeto con el resumen financiero y las transacciones de cada equipo.
     */
    _processAndSummarizeTransactions(transactions, teamIds) {
        console.log("Procesando y resumiendo transacciones por usuario...");
        const summaryData = {};
        // Inicializamos la estructura para cada equipo
        for (const teamId of teamIds) {
            summaryData[teamId] = {
                totalIngresos: 0,
                totalGastos: 0,
                transacciones: []
            };
        }

        for (const tx of transactions) {
            if (!this.allowedTypes.has(tx.type) || !tx.content) continue;

            // Usamos un Set para no añadir la misma transacción varias veces a un usuario
            const usersInThisTx = new Set();

            switch (tx.type) {
                case 'market':
                    for (const item of tx.content) {
                        if (item.to?.id && teamIds.has(item.to.id)) {
                            const userId = item.to.id;
                            summaryData[userId].totalGastos += item.amount;
                            usersInThisTx.add(userId);
                        }
                    }
                    break;

                case 'transfer':
                    for (const item of tx.content) {
                        // Ingresos para el vendedor
                        if (item.from?.id && teamIds.has(item.from.id)) {
                            summaryData[item.from.id].totalIngresos += item.amount;
                            usersInThisTx.add(item.from.id);
                        }
                        // Gastos para el comprador (si existe y es un usuario de la liga)
                        if (item.to?.id && teamIds.has(item.to.id) && item.from) {
                            summaryData[item.to.id].totalGastos += item.amount;
                            usersInThisTx.add(item.to.id);
                        }
                    }
                    break;

                case 'adminTransfer':
                    for (const item of tx.content) {
                        // Un 'adminTransfer' a un usuario ('to') es un gasto para ese usuario (ej. para cancelar una venta errónea)
                        if (item.to?.id && teamIds.has(item.to.id)) {
                            summaryData[item.to.id].totalGastos += item.amount;
                            usersInThisTx.add(item.to.id);
                        }
                        // Un 'adminTransfer' desde un usuario ('from') es un ingreso para ese usuario (ej. para cancelar una compra errónea)
                        if (item.from?.id && teamIds.has(item.from.id)) {
                            summaryData[item.from.id].totalIngresos += item.amount;
                            usersInThisTx.add(item.from.id);
                        }
                    }
                    break;

                case 'clauseIncrement':
                    for (const item of tx.content) {
                        if (item.user?.id && teamIds.has(item.user.id)) {
                            const userId = item.user.id;
                            const amount = item.amount;
                            // Si el 'amount' es negativo es un ingreso, si es positivo es un gasto.
                            amount < 0 ? summaryData[userId].totalIngresos += Math.abs(amount) : summaryData[userId].totalGastos += amount;
                            usersInThisTx.add(userId);
                        }
                    }
                    break;

                case 'roundFinished':
                    // Los bonus de la jornada son ingresos para los usuarios.
                    for (const result of tx.content.results) {
                        if (result.user?.id && teamIds.has(result.user.id)) {
                            const userId = result.user.id;
                            summaryData[userId].totalIngresos += result.bonus;
                            usersInThisTx.add(userId);
                        }
                    }
                    break;
            }

            // Añadimos la transacción completa a la lista de cada usuario involucrado
            for (const userId of usersInThisTx) {
                summaryData[userId].transacciones.push(tx);
            }
        }
        return summaryData;
    }

    /** Guarda los datos agrupados en el fichero JSON de salida. */
    async _saveResults(data) {
        try {
            const jsonData = JSON.stringify(data, null, 4);
            await fs.writeFile(this.outputFile, jsonData, 'utf-8');
            console.log(`Resultados guardados exitosamente en '${this.outputFile}'`);
        } catch (error) {
            console.error(`Error al escribir en el fichero de salida:`, error);
        }
    }

    /** Ejecuta el proceso completo. */
    async run() {
        const teamIds = await this._loadTeamIds();
        if (!teamIds) return;

        const transactions = await this._fetchAllTransactions();
        if (!transactions) return;

        // Reemplazamos la llamada al método antiguo
        const summary = this._processAndSummarizeTransactions(transactions, teamIds);
        await this._saveResults(summary);
    }
}

File: balance_updater.py
import requests
import json
import os

class BalanceUpdater:
    """
    Una clase para obtener los movimientos del tablón de una liga de Biwenger,
    filtrarlos y agruparlos por usuario.
    """

    def __init__(self, league_id, user_id, token, input_file, output_file):
        """
        Inicializa el actualizador de balances.

        Args:
            league_id (int): El ID de la liga.
            user_id (int): El ID del usuario para la autenticación.
            token (str): El token de autorización (Bearer).
            input_file (str): Ruta al JSON de equipos.
            output_file (str): Ruta donde se guardará el JSON resultante.
        """
        self.api_url = f"https://biwenger.as.com/api/v2/league/{league_id}/board"
        self.headers = {
            'x-league': str(league_id),
            'x-user': str(user_id),
            'x-version': '628',
            'Authorization': f'Bearer {token}'
        }
        self.input_file = input_file
        self.output_file = output_file
        self.allowed_types = {"transfer", "market", "clauseIncrement"}

    def _load_team_ids(self):
        """Carga los IDs de los equipos desde el fichero JSON de entrada."""
        try:
            with open(self.input_file, 'r', encoding='utf-8') as f:
                teams = json.load(f)
            # Usamos un set para una búsqueda más eficiente
            return {team['id'] for team in teams}
        except FileNotFoundError:
            print(f"Error: El fichero de entrada no fue encontrado en '{self.input_file}'")
            return set()
        except json.JSONDecodeError:
            print(f"Error: El fichero '{self.input_file}' no es un JSON válido.")
            return set()

    def _fetch_all_transactions(self):
        """
        Obtiene todas las transacciones del tablón de anuncios de la API,
        manejando la paginación.
        """
        all_transactions = []
        offset = 0
        limit = 100  # Límite por página

        print("Iniciando la obtención de transacciones de la API...")
        while True:
            params = {'offset': offset, 'limit': limit}
            try:
                response = requests.get(self.api_url, headers=self.headers, params=params)
                response.raise_for_status()  # Lanza un error para respuestas 4xx/5xx
                
                data = response.json().get('data', [])
                if not data:
                    print("No se encontraron más transacciones. Finalizando paginación.")
                    break
                
                all_transactions.extend(data)
                print(f"Obtenidas {len(data)} transacciones. Total acumulado: {len(all_transactions)}")
                
                # Si la API devuelve menos resultados que el límite, es la última página.
                if len(data) < limit:
                    break

                offset += limit

            except requests.exceptions.RequestException as e:
                print(f"Error al contactar la API de Biwenger: {e}")
                return None

        return all_transactions

    def _group_transactions(self, transactions, team_ids):
        """Filtra y agrupa las transacciones por ID de usuario."""
        print("Agrupando transacciones por usuario...")
        grouped_data = {team_id: [] for team_id in team_ids}

        for tx in transactions:
            tx_type = tx.get('type')
            user_info = tx.get('user')
            
            if user_info and tx_type in self.allowed_types:
                user_id = user_info.get('id')
                if user_id in grouped_data:
                    grouped_data[user_id].append(tx)
        
        return grouped_data

    def _save_results(self, data):
        """Guarda los datos agrupados en el fichero JSON de salida."""
        try:
            with open(self.output_file, 'w', encoding='utf-8') as f:
                json.dump(data, f, ensure_ascii=False, indent=4)
            print(f"Resultados guardados exitosamente en '{self.output_file}'")
        except IOError as e:
            print(f"Error al escribir en el fichero de salida: {e}")

    def run(self):
        """Ejecuta el proceso completo."""
        team_ids = self._load_team_ids()
        if not team_ids:
            return

        transactions = self._fetch_all_transactions()
        if transactions is None:
            return

        grouped_transactions = self._group_transactions(transactions, team_ids)
        self._save_results(grouped_transactions)

if __name__ == '__main__':
    # --- CONFIGURACIÓN ---
    LEAGUE_ID = 1683930
    USER_ID = 10352455
    # ¡IMPORTANTE! Este token caduca. Deberás obtener uno nuevo si falla la autenticación.
    AUTH_TOKEN = 'eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOjI3OTU3MTkwLCJpYXQiOjE3NTM4ODU3NjJ9.paAgVghdKnd284Eth572z9dGjT2j_FduqmLHi0qKiXk'
    INPUT_JSON_PATH = os.path.join(os.path.dirname(__file__), 'equipos.json')
    OUTPUT_JSON_PATH = os.path.join(os.path.dirname(__file__), 'transacciones_agrupadas.json')

    updater = BalanceUpdater(LEAGUE_ID, USER_ID, AUTH_TOKEN, INPUT_JSON_PATH, OUTPUT_JSON_PATH)
    updater.run()